src/crypto/CHIPCryptoPALmbedTLS.cpp

```diff
@@ -1207,7 +1209,50 @@ CHIP_ERROR GetNumberOfCertsFromPKCS7(const char * pkcs7, uint32_t * n_certs)
 CHIP_ERROR ValidateCertificateChain(const uint8_t * rootCertificate, size_t rootCertificateLen, const uint8_t * caCertificate,
                                     size_t caCertificateLen, const uint8_t * leafCertificate, size_t leafCertificateLen)
 {
-    return CHIP_ERROR_NOT_IMPLEMENTED;
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    mbedtls_x509_crt mbed_cert;
+    mbedtls_x509_crt root_cert;
+    int result;
+    uint32_t flags;
+
+    mbedtls_x509_crt_init(&mbed_cert);
+    mbedtls_x509_crt_init(&root_cert);
+
+    /* Start of chain  */
+    result = mbedtls_x509_crt_parse(&mbed_cert, Uint8::to_const_uchar(leafCertificate), leafCertificateLen);
+    VerifyOrExit(result == 0, error = CHIP_ERROR_INTERNAL);
+    /* Add the intermediate to the chain  */
+    result = mbedtls_x509_crt_parse(&mbed_cert, Uint8::to_const_uchar(caCertificate), caCertificateLen);
+    VerifyOrExit(result == 0, error = CHIP_ERROR_INTERNAL);
+    /* Add the root to the chain */
+    result = mbedtls_x509_crt_parse(&mbed_cert, Uint8::to_const_uchar(rootCertificate), rootCertificateLen);
+    VerifyOrExit(result == 0, error = CHIP_ERROR_INTERNAL);
+
+    /* Parse the root cert */
+    result = mbedtls_x509_crt_parse(&root_cert, Uint8::to_const_uchar(rootCertificate), rootCertificateLen);
+    VerifyOrExit(result == 0, error = CHIP_ERROR_INTERNAL);
+
+    /* Verify the chain against the root */
+    result = mbedtls_x509_crt_verify(&mbed_cert, &root_cert, NULL, NULL, &flags, NULL, NULL);
+    VerifyOrExit(result == 0 && flags == 0, error = CHIP_ERROR_CERT_NOT_TRUSTED);
```

Comment:
Would it be possible to have more errors for at least a few common cases? This can be done as a follow-up.

Reply:
I think so - will mark it as TODO